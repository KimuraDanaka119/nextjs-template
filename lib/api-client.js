import fetch from 'isomorphic-unfetch';
import config from 'configs/config';

export default class ApiClient {
  async doFetch(path) {
    const url = config.api.baseURL + path;

    const res = await fetch(url);
    const data = await res.json();

    return {
      statusCode: res.status,
      data,
    };
  }

  async getShows(keyword) {
    const res = await this.doFetch(`/search/shows?q=${keyword}`);

    return res;
  }

  async getShow(id) {
    const res = await this.doFetch(`/shows/${id}`);

    return res;
  }
}
