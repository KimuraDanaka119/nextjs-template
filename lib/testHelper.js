import ApiClient from './api-client';

function t(key) {
  return key;
}

const config = {
  env: 'test',
  versionHash: 'test',
  paths: {
    static: '/static/test',
    styles: '/static/test/styles',
    images: '/static/test/images',
    locales: '/static/test/locales',
  },
  api: {
    baseURL: '.',
  },
};

const apiClient = new ApiClient('.');

export { t, config, apiClient };
