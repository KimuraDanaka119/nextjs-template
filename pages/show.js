import React from 'react';
import PropTypes from 'prop-types';

import withPage from '../components/hoc/withPage';
import withI18N from '../components/hoc/withI18N';
import Layout from '../components/common/Layout';
import Meta from '../components/common/Meta';
import Error from '../components/common/Error';
import Show from '../components/content/Show';

import ApiClient from '../lib/api-client';

class ShowPage extends React.Component {
  static async getInitialProps({ query, config }) {
    const { id } = query;
    const apiClient = new ApiClient(config.api.baseURL);
    const result = await apiClient.getShow(id);

    return { result };
  }

  render() {
    const { t, result } = this.props;

    return (
      <Layout>
        <Meta
          title={t('show:meta.title', { name: result.data.name })}
          description={t('show:meta.description')}
        />
        {result.statusCode >= 400
          ? <Error statusCode={result.statusCode} />
          : <Show result={result} />
        }
      </Layout>
    );
  }
}

ShowPage.propTypes = {
  result: PropTypes.object,
  t: PropTypes.func,
};

export default withPage(withI18N(ShowPage, ['show']));
export const undecorated = ShowPage;
