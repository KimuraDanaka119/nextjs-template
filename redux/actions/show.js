import ApiClient from 'lib/api-client';

const apiClient = new ApiClient();

export const showHasErrored = bool => ({
  type: 'SHOW_HAS_ERRORED',
  hasErrored: bool,
});

export const showIsLoading = bool => ({
  type: 'SHOW_IS_LOADING',
  isLoading: bool,
});

export const showFetchDataSuccess = show => ({
  type: 'SHOW_FETCH_DATA_SUCCESS',
  show,
});

export const showFetchData = id => (dispatch) => {
  dispatch(showIsLoading(true));

  return apiClient.getShow(id).then((res) => {
    dispatch(showIsLoading(false));
    dispatch(showFetchDataSuccess(res.data));
  });
};
