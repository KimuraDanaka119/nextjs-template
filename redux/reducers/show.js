const showHasErrored = (state = false, action) => {
  switch (action.type) {
  case 'SHOW_HAS_ERRORED':
    return action.hasErrored;

  default:
    return state;
  }
};

const showIsLoading = (state = false, action) => {
  switch (action.type) {
  case 'SHOW_IS_LOADING':
    return action.isLoading;

  default:
    return state;
  }
};

const show = (state = {}, action) => {
  switch (action.type) {
  case 'SHOW_FETCH_DATA_SUCCESS':
    return action.show;

  default:
    return state;
  }
};

export default {
  showHasErrored,
  showIsLoading,
  show,
};
