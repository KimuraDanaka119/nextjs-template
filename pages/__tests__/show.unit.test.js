import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import { undecorated as ShowPage } from '../show';
import { t, config, apiClient } from '../../lib/testHelper';

describe('<ShowPage />', () => {
  it('getInitialProps should return object with status code', async () => {
    let props = await ShowPage.getInitialProps({ query: { id: '975' }, config });

    expect(props).toBeDefined();
    expect(props).toBeDefined();
    expect(props.result.statusCode).toEqual(200);
    expect(props.result.data.id).toEqual(975);

    props = await ShowPage.getInitialProps({ query: { id: '123456789' }, config });

    expect(props).toBeDefined();
    expect(props).toBeDefined();
    expect(props).toBeDefined();
    expect(props.result.statusCode).toEqual(404);
  });

  it('should render correctly with searchable result', async () => {
    const result = await apiClient.getShow('975');

    const show = shallow(<ShowPage t={t} result={result} />);
    expect(toJson(show)).toMatchSnapshot();
  });

  it('should render correctly with non-searchable result', async () => {
    const result = await apiClient.getShow('123456789');

    const show = shallow(<ShowPage t={t} result={result} />);
    expect(toJson(show)).toMatchSnapshot();
  });
});
