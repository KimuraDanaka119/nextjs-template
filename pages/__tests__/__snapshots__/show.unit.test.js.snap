// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShowPage /> should render correctly with non-searchable result 1`] = `
<Layout>
  <Meta
    description="show:meta.description"
    title="show:meta.title"
  />
  <Translate(I18NComponent)
    statusCode={404}
  />
</Layout>
`;

exports[`<ShowPage /> should render correctly with searchable result 1`] = `
<Layout>
  <Meta
    description="show:meta.description"
    title="show:meta.title"
  />
  <Translate(I18NComponent)
    result={
      Object {
        "data": Object {
          "_links": Object {
            "previousepisode": Object {
              "href": "http://api.tvmaze.com/episodes/95649",
            },
            "self": Object {
              "href": "http://api.tvmaze.com/shows/975",
            },
          },
          "externals": Object {
            "imdb": "tt0059968",
            "thetvdb": 77871,
            "tvrage": 2719,
          },
          "genres": Array [
            "Comedy",
            "Action",
            "Children",
            "Crime",
            "Science-Fiction",
          ],
          "id": 975,
          "image": Object {
            "medium": "http://static.tvmaze.com/uploads/images/medium_portrait/6/16463.jpg",
            "original": "http://static.tvmaze.com/uploads/images/original_untouched/6/16463.jpg",
          },
          "language": "English",
          "name": "Batman",
          "network": Object {
            "country": Object {
              "code": "US",
              "name": "United States",
              "timezone": "America/New_York",
            },
            "id": 3,
            "name": "ABC",
          },
          "officialSite": null,
          "premiered": "1966-01-12",
          "rating": Object {
            "average": 8.2,
          },
          "runtime": 30,
          "schedule": Object {
            "days": Array [
              "Thursday",
            ],
            "time": "19:30",
          },
          "status": "Ended",
          "summary": "<p>Wealthy entrepreneur Bruce Wayne and his ward Dick Grayson lead a double life: they are actually crime fighting duo Batman and Robin. A secret Batpole in the Wayne mansion leads to the Batcave, where Police Commissioner Gordon often calls with the latest emergency threatening Gotham City. Racing to the scene of the crime in the Batmobile, Batman and Robin must (with the help of their trusty Bat-utility-belt) thwart the efforts of a variety of master criminals, including Catwoman, Egghead, The Joker, King Tut, The Penguin, and The Riddler.</p>",
          "type": "Scripted",
          "updated": 1512931633,
          "url": "http://www.tvmaze.com/shows/975/batman",
          "webChannel": null,
          "weight": 79,
        },
        "statusCode": 200,
      }
    }
  />
</Layout>
`;
